### PR TITLE
Custom embeddings

### PR DESCRIPTION
**Pull request by x-tabdeveloping**
Added custom embedding functionality to E5 and Cohere models.

### CODE REVIEW

**Comment by x-tabdeveloping** [comment]
Addresses #75 and #76

File: src/seb/registered_models/cohere_models.py [hunk 1]
```diff
@@ -4,54 +4,74 @@
...
+        *,
+        task: Task,
+        **kwargs: Any,  # noqa: ARG002
+    ) -> torch.Tensor:
+        if task.task_type == "Classification":
+            input_type = "classification"
+        elif task.task_type == "Clustering":
+            input_type = "clustering"
+        else:
+            input_type = "search_document"
```
Comment (KennethEnevoldsen) [inline src/seb/registered_models/cohere_models.py:52]: retrieval should be search document. 

Should it be the default as well?

**Review by KennethEnevoldsen — CHANGES_REQUESTED** [review]
We should run at least the smallest e5 model to validate.

We should also remove the cache for the changed models (all e5 models except mistral and cohere)

File: src/seb/registered_models/cohere_models.py (hunk 1 above)
Comment (x-tabdeveloping) [reply src/seb/registered_models/cohere_models.py:52]: Retrieval tasks use asymmetric encoding anyways, so hopefully this function will never run in that case. As for the default, I don't see why not. The two tasks that do not have specific input types in Cohere are STS and BitextMining, both of which (in my humble judgement) are most similar to retrieval. What do you think, @KennethEnevoldsen?

Comment (KennethEnevoldsen) [reply src/seb/registered_models/cohere_models.py:52]: I think this is fine. Just wanted to make sure we had considered it.

**Comment by x-tabdeveloping** [comment]
Hmm, seems like the task encodes passages with the normal `encode()` but encodes queries with `encode_query()` which is not good, because normal `encode()` also does the `query: ` prefixing. I wonder if this is the intended behaviour and I just misunderstood, or shouldn't `encode_document()` run when encoding passages? @Muennighoff

**Comment by x-tabdeveloping** [comment]
Oh no, I'm wrong, only the `encode()` method is being run for some reason

**Comment by x-tabdeveloping** [comment]
I think I see why, I will try to fix it.

File: src/seb/registered_models/cohere_models.py (hunk 1 above)
Comment (KennethEnevoldsen) [reply src/seb/registered_models/cohere_models.py:52]: No problem, just wanted to make sure

**Comment by KennethEnevoldsen** [comment]
implementing it for the seb.EmbeddingModel seems very reasonable. 

I was very unsure about the MTEBTaskModel implementation. An alternative solution I considered was to add a "set_task()" method to the seb.EmbeddingModel.

**Comment by x-tabdeveloping** [comment]
Yeah I really don't like MTEBTaskModel, I will think a bit about a possible solution. Would there be some circular shenanigans if the model had a task from the getgo?

**Comment by x-tabdeveloping** [comment]
Or in other words: Would it be the end of the world if `EmbeddingModel` had a task at initialisation?

**Comment by x-tabdeveloping** [comment]
@KennethEnevoldsen Problem: `encode()` always takes a `Task`, but `encode_queries()` and `encode_corpus()` should be able to get away without it, cause MTEB doesn't pass `Task` along to these. The issue is that by default (if we make a default implementation for `encode_queries()` in `EmbeddingModel`) both of the new methods have to be able to call `encode()`, but there is no task we can pass to `encode()`.

What do we do?
1. Make `task: Optional[Task] = None` all places
2. Make an empty task that has `task_type = "Retrieval"` and pass that to `encode()` when it's being called from `encode_queries()` or `encode_corpus()`

**Comment by x-tabdeveloping** [comment]
Another option would be to propagate an `AttributeError` from the `EmbeddingModel._model` if it doesn't have an `encode_query()`, the problem with this is that MTEB checks for the function with `getattr()`, not a try-catch block, which would pass because `EmbeddingModel` has both attributes.

**Comment by x-tabdeveloping** [comment]
Yet another thing: I'm not entirely comfortable with having a protocol named `Encoder` and then a class called `EmbeddingModel`. It confused me at first what these two were doing. It would be nice to communicate somehow that `EmbeddingModel` lazy loads another model when `encode()` is called, and carries model metadata.

**Comment by x-tabdeveloping** [comment]
We could also potentially fix the task situation by making all `Encoder`-s have a task attribute.
Then instead of `EmbeddingModel` we would have an `ModelLoader`:
```python
class Encoder(Protocol):
     task: Task
     
     def encode(self, sentences: list[str], **kwargs):
          pass
    # And then optionally the other two methods

@dataclass
class ModelLoader:
    meta: ModelMeta
    loader: Callable[[Task], Encoder]
    
class Task(Protocol):
     def evaluate(self, loader: ModelLoader) -> TaskResult:
          model = loader(self)
          ...
```

**Comment by KennethEnevoldsen** [comment]
This is outside the scope of this PR. I have created an issue instead.

**Comment by x-tabdeveloping** [comment]
@KennethEnevoldsen do you approve?

**Comment by KennethEnevoldsen** [comment]
> @KennethEnevoldsen do you approve?

You can just re-request a review - then I will be notified

File: src/seb/interfaces/model.py
Comment (KennethEnevoldsen) [inline src/seb/interfaces/model.py:1]: Looks good

File: src/seb/interfaces/mteb_task.py [hunk 2]
```diff
@@ -76,8 +68,12 @@ def get_descriptive_stats(self) -> DescriptiveDatasetStats:
...
     def evaluate(self, model: Encoder) -> TaskResult:
         split = self.mteb_task.description["eval_splits"][0]
-        task_model = MTEBTaskModel(model, self)
-        scores = self.mteb_task.evaluate(task_model, split=split)
+        # Infusing task into encode()
+        original_encode = model.encode
+        model.encode = partial(model.encode, task=self)
+        scores = self.mteb_task.evaluate(model, split=split)
+        # Resetting encode to original
+        model.encode = original_encode
```
Comment (KennethEnevoldsen) [inline src/seb/interfaces/mteb_task.py:76]: If an error is raised here would get us some odd problems. One solution is to make a context manager another is to:

```suggestion
        # Infusing task into encode()
        original_encode = model.encode
        model.encode = partial(model.encode, task=self)
        try:
          scores = self.mteb_task.evaluate(model, split=split)
          model.encode = original_encode # reset encode function
        except Exception as e:
          # this ensure that models encode function is always reset
          model.encode = original_encode
          raise(e)

```

File: src/seb/registered_tasks/multilingual.py [hunk 3]
```diff
@@ -96,11 +93,13 @@ def get_descriptive_stats(self) -> DescriptiveDatasetStats:
...
             scores = {}
-            _model = MTEBTaskModel(model, self)
+            original_encode = model.encode
+            model.encode = partial(model.encode, task=self)
             for lang, mteb_task in self.mteb_tasks.items():
                 mteb_task.load_data()
-                score = mteb_task.evaluate(_model)
+                score = mteb_task.evaluate(model)
                 scores[lang] = score
+            model.encode = original_encode
```
Comment (KennethEnevoldsen) [inline src/seb/registered_tasks/multilingual.py:102]: Same as above

**Review by KennethEnevoldsen — APPROVED** [review]
One suggestion otherwise I believe it looks good

File: src/seb/interfaces/mteb_task.py (hunk 2 above)
Comment (x-tabdeveloping) [reply src/seb/interfaces/mteb_task.py:76]: Ahhh smart, I haven't even thought of that, thanks Kenneth

**Pull request merged**